### ISSUE TITLE
Task Manager Auto Start Refeature

### ISSUE THREAD

**Issue by shinra358** [Ryochan7/DS4Windows#16]
I would like to ask for the check box for the autostart option to create a task manager entry and after the exe is started again (or, the shortcut exe set in windows auto start folder that calls to the task manager option set by the program), it will call for the task manager option so that it will always start in admin mode. This would be helpful for MMOs that require admin access for the controllers to work. If unchecked, it would delete said task manager option and windows folder autostart shortcut.

**Comment by pralima87** [comment]
Is really easy to create the task yourself man:

1-Create the task: Run DS4Windows
-Select "Run only when user is logged on"
-Trick "Run with highest privileges"
-Trick "Hidden"

2-Add trigger
-Begin the task: "At log on"
-Trick "enable"

3-Create the Action to launch DS4W
-Add "new" action: "Start a Program"
-Program/script: "%windir%\System32\cmd.exe" (No quotes)
-Add arguments: /c start "Run DS4Windows" "[C:.......]\DS4Windows.exe"*

OBS: Replace all the "[C:.......\DS4\DS4Tool - J2K]" with the path to your DS4Windows.exe file.
Last make sure DS4Windows.exe is already with Admin privileges.

Done.

**Comment by shinra358** [comment]
Yes, I already know how to do it. This is what I've explained. I do it with a program that I coded in ahk and the program does everything automatically. But the focal point is making the program as self reliant as possible. If it is a feature, why should it rely on something else ya know. I am the person that requested this original feature but at the time I didn't have a game that used admin priviledges. Now I do. If you'd like the winapi code I used, id be happy to link you to it so that you wouldn't have to waste time coming up with it yourself and it could be converted and integrated into this project.

**Comment by Ryochan7** [comment]
I will keep this in mind. The only part that might be a problem is making sure that the task will launch with administrator privileges. If you already have some reference code, you can post a Gist here and that will make getting the feature integrated that much easier.

**Comment by shinra358** [comment]
This is the code made my skan with a tiny bit of changes by myself (not much):
*call variable would be: RunAsTask()

```
RunAsTask()                          ;  By SKAN,  http://goo.gl/yG6A1F,  CD:19/Aug/2014 | MD:22/Aug/2014
	{
 
		Local CmdLine, TaskName, TaskExists, XML, TaskSchd, TaskRoot, RunAsTask
		Local TASK_CREATE := 0x2,  TASK_LOGON_INTERACTIVE_TOKEN := 3 
		 
		Try TaskSchd  := ComObjCreate( "Schedule.Service" ),    TaskSchd.Connect()
		, TaskRoot  := TaskSchd.GetFolder( "\" )
		Catch
			Return "", ErrorLevel := 1    
;		 
		SplitPath, Main, , , , A_Main
		CmdLine       := ( A_IsCompiled ? "" : """"  A_AhkPath """" A_Space ) ( """" A_ScriptFullpath """"  )
		TaskName      := "Enhancer [" A_Main "]" 
;		 
		Try RunAsTask := TaskRoot.GetTask( TaskName )
		TaskExists    := ! A_LastError 
;		 		 
		If ( not A_IsAdmin and TaskExists )      
			{ 			 
				RunAsTask.Run( "" )
				ExitApp			 
			}
;		 
		If ( not A_IsAdmin and not TaskExists )  
			{ 			 
				Run *RunAs %CmdLine%, %A_ScriptDir%, UseErrorLevel
				ExitApp			 
			}
;		 
		If ( A_IsAdmin and not TaskExists )      
			{  
;			 
				XML := "
				( LTrim Join
					<?xml version=""1.0"" ?><Task xmlns=""http://schemas.microsoft.com/windows/2004/02/mit/task""><Regi
					strationInfo /><Triggers /><Principals><Principal id=""Author""><LogonType>InteractiveToken</LogonT
					ype><RunLevel>HighestAvailable</RunLevel></Principal></Principals><Settings><MultipleInstancesPolic
					y>IgnoreNew</MultipleInstancesPolicy><DisallowStartIfOnBatteries>false</DisallowStartIfOnBatteries><
					StopIfGoingOnBatteries>false</StopIfGoingOnBatteries><AllowHardTerminate>false</AllowHardTerminate>
					<StartWhenAvailable>false</StartWhenAvailable><RunOnlyIfNetworkAvailable>false</RunOnlyIfNetworkAva
					ilable><IdleSettings><StopOnIdleEnd>true</StopOnIdleEnd><RestartOnIdle>false</RestartOnIdle></IdleS
					ettings><AllowStartOnDemand>true</AllowStartOnDemand><Enabled>true</Enabled><Hidden>false</Hidden><
					RunOnlyIfIdle>false</RunOnlyIfIdle><DisallowStartOnRemoteAppSession>false</DisallowStartOnRemoteApp
					Session><UseUnifiedSchedulingEngine>false</UseUnifiedSchedulingEngine><WakeToRun>false</WakeToRun><
					ExecutionTimeLimit>PT0S</ExecutionTimeLimit></Settings><Actions Context=""Author""><Exec>
					<Command>"   (  A_IsCompiled ? A_ScriptFullpath : A_AhkPath )       "</Command>
					<Arguments></Arguments>
					</Exec></Actions></Task>
				)"    
;				 
				TaskRoot.RegisterTask( TaskName, XML, TASK_CREATE, "", "", TASK_LOGON_INTERACTIVE_TOKEN )			 
			}         
;		 
		Return TaskName, ErrorLevel := 0
	}
```

Of course the taskmanager descriptions and start location would need to be edited to suit this program. This will create the task.

You would have to create code (if/else statement) that says "if this has already been done, and the start location hasn't been changed, then initiate the task." uac will only prompt once and after that, one would be good to go. 

Double clicking the exe to initate the program would automatically start the task instead of run the exe (when you code that in of course). The autostart shortcut would be placed in the startup folder like normal but with the task implementations so that it will start with admin priviledges. 

Of course this still has ahk variables in it so similar things would have to be converted to fit the current language. But most of it is pretty much winapi.

**Comment by Ryochan7** [comment]
Thanks for posting the example code. For now, I am testing creating the task using the Task Scheduler Managed Wrapper library. I have gotten a task created when the "Run at Startup" checkbox is checked and the task is deleted when unchecked. The task works as expected when launched from the shortcut and DS4Windows will run with elevated privileges when activated. The only problem right now is that the Startup entry in Task Manager shows up as "Task Scheduler Configuration Tool" since I currently am using schtasks to run the task. On the plus side, no UAC prompt appears. I will have to try this as a startup mode for DS4Windows similar to Exclusive Mode activation and see if it still works with no UAC prompt.

http://taskscheduler.codeplex.com/

**Comment by shinra358** [comment]
Excellent. Don't forget to code in the instance where a person moves ds4windows before unchecking to a different directory, that the next time it is ran, the old location task schedule is erased before new location version is put in. Do this by checking for the name of the task, then grab the old task ds4 location and then grab the new location, compair, and then get ds4 to delete the location where it is no longer at. ;)

**Comment by Ryochan7** [comment]
I am about ready to give up on this. It is unfortunate since I got so close to getting things working. The problem is that DS4Windows takes a long time to fully start when launched from a task at startup no matter which method or settings for the task I use. DS4Windows is shown in Task Manager immediately but the thread count is very low to start and increases over time until the main window finally appears on screen. This problem is not isolated to DS4Windows since the same issues come up when I have the task set to launch other programs at log on as well. However, once the system is fully loaded, the program associated with a task will launch instantly.

**Comment by shinra358** [comment]
Well, the thing about that is, no one is using the controller immediately when windows starts because we mostly have to wait for everything else to load before we can do anything. No one is immediately playing a game at this time either. So it not coming up as quickly as youd like is probably normal and wouldn't interfere with anything. I know I have to wait 2 to 3 minutes before I'm even able to move the mouse lol.

**Comment by Ryochan7** [comment]
It is a shame since the normal startup routine takes less than 10 seconds to fully launch. Unfortunately there is no way to force the application to run as the administrator directly using that method. Even if the runas workaround were used, a UAC window would pop up when starting the new instance of the program. At least the task method bypasses that problem and launches as the administrator with no interference.

I guess I will have the task method be optional but the current behavior will be the default method used for startup.

**Comment by shinra358** [comment]
Sounds good to me ;)

**Comment by Ryochan7** [comment]
The newest version has this feature implemented.

https://github.com/Ryochan7/DS4Windows/releases/download/v1.4.62/DS4Windows_1.4.62_x64.zip

**Comment by shinra358** [comment]
Thx, appreciate it. I'll let you know if I run into anything ;)

**Comment by shinra358** [comment]
Every time I try to play Blade & Soul and I start up the game (after ds4 has already started at windows start up), ds4 silently crashes. Maybe the theres a task option to start if program fails or something? And also when I double click it normally, it doesn't call for the task scheduled task, but starts in non admin mode.

**Comment by Ryochan7** [comment]
Double clicking the exe should still launch the application as a non-admin. There is always the option to have the application always launch as the administrator in Explorer although you will have to interact with a UAC prompt. If you really want to run DS4Windows via a task outside of startup, you can either run the startup shortcut or go to Control Panel > System and Security > Administrative Tools > Task Scheduler and run the `RunDS4Windows` task from there; the startup shortcuts location can be accessed by running `shell:startup` using the run dialog window.

As for the Blade & Soul problem, there could be a number of reasons for that problem. I have registered an account and I am downloading the game now. This situation will have to be tested before I can have an idea about what needs to be changed.

**Comment by Ryochan7** [comment]
The log that you posted did show a potential problem that I have corrected. Hopefully that is the only issue and the game will run fine once it finishes downloading.

**Comment by Ryochan7** [comment]
Just tried this out for a few startups with the previous commit and I didn't run into that problem. Blade and Soul works fine when using either KB+M controls or Xbox 360 gamepad controls.

**Comment by Ryochan7** [comment]
Please try the latest release

https://github.com/Ryochan7/DS4Windows/releases/download/v1.4.63/DS4Windows_1.4.63_x64.zip

**Comment by shinra358** [comment]
Looks good so far. As far as if it crashes again while testing, to speed up the process of restarting the program, could you then make the program create the same shortcut to the ds4 folder and if it is on the system drive, create that shortcut to the desktop? That way we can continue testing without interruption (would only be created if the start with windows box is checked)

**Comment by Ryochan7** [comment]
For now, you have to make sure to run DS4Windows as the admin when you want to enable the option. The latest release now has the option disabled if you are not running as an admin.

Adding a desktop shortcut option is not really necessary. You can create a desktop shortcut from the startup shortcut if you want a desktop shortcut with the runtask option added.

DS4Windows does check the shortcut target against the current location of the program when launched. If the two differ then DS4Windows will overwrite the startup shortcut with one that includes the new program location. The old task will get deleted and a new task with the new program location will be created.

Lastly, even when using the Hidden option for tasks, the cmdbox will still show up. I have not found a way to keep the cmdbox hidden while launching DS4Windows. cmd.exe is being used to launch DS4Windows so that the task only monitors the execution of cmd.exe instead of the instance of DS4Windows. The task will only be activate for the duration that cmd.exe is running rather than staying active waiting for DS4Windows to quit.

**Comment by shinra358** [comment]
hide console solutions: http://stackoverflow.com/questions/836427/how-to-run-a-c-sharp-console-application-with-the-console-hidden

offset console window offscreen by using a huge x and/or y value: http://stackoverflow.com/questions/1548838/setting-position-of-a-console-window-opened-in-a-winforms-app

**Comment by Ryochan7** [comment]
A cmd window being shown is behavior associated with the task level and not from anything that can be changed in .NET. I cannot find a way to make the cmd window not appear when using the task action as it is. The current action is the cleanest way that I have seen to make the task work. The issue is minor so this should probably just be expected behavior for this kind of hack.
